src/incubator/TextField/index.tsx

```diff
@@ -0,0 +1,148 @@
+import React, {useMemo} from 'react';
+import {ViewStyle, TextStyle} from 'react-native';
+import {
+  asBaseComponent,
+  forwardRef,
+  ForwardRefInjectedProps
+} from '../../commons/new';
+import View from '../../components/view';
+import {ButtonPropTypes} from '../../components/button';
+import {ValidationMessagePosition} from './types';
+import Input, {InputProps} from './Input';
+import AccessoryButton from './AccessoryButton';
+import ValidationMessage, {ValidationMessageProps} from './ValidationMessage';
+import Label, {LabelProps} from './Label';
+import FieldContext from './FieldContext';
+import withFieldState, {
+  FieldStateInjectedProps,
+  FieldStateProps
+} from './withFieldState';
+import FloatingPlaceholder, {
+  FloatingPlaceholderProps
+} from './FloatingPlaceholder';
+import CharCounter, {CharCounterProps} from './CharCounter';
+
+interface TextFieldProps
+  extends InputProps,
+    LabelProps,
+    FloatingPlaceholderProps,
+    FieldStateProps,
+    ValidationMessageProps,
+    CharCounterProps {
+  leadingButton?: ButtonPropTypes;
+  trailingButton?: ButtonPropTypes;
+  floatingPlaceholder?: boolean;
+  floatingPlaceholderStyle?: TextStyle;
+  validationMessagePosition?: ValidationMessagePosition;
+  fieldStyle?: ViewStyle;
+  containerStyle?: ViewStyle;
+}
+
+interface InternalTextFieldProps
+  extends TextFieldProps,
+    Omit<FieldStateInjectedProps, keyof InputProps>,
+    ForwardRefInjectedProps {}
+
+interface StaticMembers {
+  validationMessagePositions: typeof ValidationMessagePosition;
+}
+
+const TextField = (
+  {
+    // General
+    fieldStyle,
+    containerStyle,
+    floatingPlaceholder,
+    floatingPlaceholderColor,
+    floatingPlaceholderStyle,
+    hint,
+    // Label
+    label,
+    labelColor,
+    labelStyle,
+    labelProps,
+    // Accessory Buttons
+    leadingButton,
+    trailingButton,
+    // Validation
+    enableErrors, // TODO: rename to enableValidation
+    validationMessage,
+    validationMessageStyle,
+    validationMessagePosition = ValidationMessagePosition.BOTTOM,
+    // Char Counter
+    showCharCounter,
+    charCounterStyle,
+    // Field State
+    fieldState,
+    // Input
+    placeholder,
+    ...props
+  }: InternalTextFieldProps
+) => {
+  const context = useMemo(() => {
+    return {...fieldState, disabled: props.editable === false};
+  }, [fieldState, props.editable]);
+
+  return (
+    <FieldContext.Provider value={context}>
+      <View style={containerStyle}>
+        <Label
```

Comment:
I believe render methods is less clean in a function components, that's why I moved to breaking each render to an actual component. 

In function component you usually need to wrap internal methods with the `useCallback` hook and I wanted to avoid that..